### ISSUE TITLE
Hold New Key Exchange Message While Asking to End Old Session

### ISSUE THREAD

**Issue by landry314** [SMSSecure/SMSSecure#29]
One thing that has always been annoying about TextSecure is if I need to redo session keys for whatever reason (bugs ruining session, new phone, etc) then I can end the session on my end and send a new key but the client on the other end is still holding the old key and the app has no clean way to deal with this. I usually have to send an unencrypted message explaining "I ended my session, please end your session and start a new one by going to the lock menu... blah blah" and have them understand what I mean and how to do it.

It would be fantastic if SMSSecure could hold the new key coming in and pull up a dialog saying "The user has sent you a new key, would you like to end your session and accept this new key?" then give the user an option to do that right there. It would be so much cleaner!

I realize this is a more serious programming task but if anyone is up for it, it would be a huge improvement in the SMSSecure key management system!

**Comment by landry314** [comment]
Yeah, I thought this would be too much to ask. SMSSecure could just be an app with keeps SMS in TextSecure and drops push OR it could develop SMS encryption. If any programmer who can do this comes along that would be beautiful.

**Comment by BastienLQ** [comment]
I don't really understand what you want. Currently, if a new key is sent over an existing secure session, the recipient will see a warning but can start the new secure session by tapping on a button. This is what you want, doesn't it?

In addition, this issue might be a duplicate of #16 about exporting keys to a new device or after a data loss.

**Comment by pR0Ps** [comment]
I just tested the following scenario on my phone (with a friend in the room)
- A has a secure session with B
- A uninstalls the app (without tearing down the session first)
- A reinstalls the app and gets a new identity key
    - At this point, if A sends a message to B, B receives the unencrypted message and displays it without a warning (the only indication is a lack of a lock icon).
    - If B sends a message to A, A gets a "Message encrypted for a non-existing session" warning and can't read the message.
- A sends a request to start a secure session to B
- B gets the request and is shown a warning that the identity has changed, taps ok.
- A has a secure session with B
- Interestingly, B can receive unencrypted messages from A before the key exchange without any big warnings (the only indication is a lack of a lock icon on the text bubble).

So in summary, the way it works now is exactly the way you want it to work?

**Comment by landry314** [comment]
i dont see any option to start a new session when one person ends the current session then tries to start it immediately. what happens is that the person who still has the old session key gets an error message or unreadable text but doesnt know why. to "refresh the keys" both parties need to end the session and then one person needs to start a new session. this all needs to be explained in unencrypted text to the person who has the old key and needs to end the session on his side as well.

**Comment by pR0Ps** [comment]
Not according to the test I just performed. When the person who thinks there's still a secure session gets another key message, they do exactly what you requested and process the key (with a warning that it's changed) and send the response.

**Comment by landry314** [comment]
Well, I remember in the past many times that I needed to tell the other person to end the session or I couldn't start a new one... but, there were many incarnations of TextSecure and it kept changing how it worked... maybe this is why.

I have not seen this warning key change dialog that it referred to above, ever. I don't know why. I will play with this more when someone is in the room with an android phone with TextSecure 2.6.4.

What I did just do is tell my friend with TS 2.6.4 to end the session and send a new key while my old session was active. What happened was I got a new session key and my app sent a new session key back and it worked. I did not get a dialog asking me if I wanted to end my old session and start a new one. I think this is a security flaw that could expose us to man-in-the-middle attacks.

Starting a new session or completely reinstalling the whole app with a new data folder or reinstalling the app without deleting the data folder are different things and will have different results.

I myself have been utterly confused by trying to downgrade from 2.7 to 2.6 by uninstalling TextSecure, deleting the /data/data/ TextSecure folder and installing 2.6 cleanly then extracting the older 2.6 data folder that i had backed up. It did not work, most of the sessions failed (most, so bizarre), it would clear all my settings or messages SOMETIMES... i have no idea how... there must be another data folder I can not find somewhere... the sessions got really screwed up.

What am I asking for?? Hmmm... I want SMSSecure to work with keys cleanly and behave as expected. I am eagerly awaiting the release on F-Droid so I can play with it. When I do I will test every possible session key mismatch, key deleting, and reinstallation problem and see what happens...

**Comment by pR0Ps** [comment]
If/when you find an issue, you can open a new issue. I'm going to close this as it seems to be resolved.
